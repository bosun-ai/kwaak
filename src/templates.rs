use anyhow::Context as _;
use anyhow::Result;
use rust_embed::Embed;
use tera::Tera;

#[derive(Embed)]
#[folder = "templates"]
pub struct Templates;

impl Templates {
    pub fn render(name: &str, context: &tera::Context) -> Result<String> {
        let byte_file =
            Templates::get(name).with_context(|| format!("Expected template {name}"))?;
        let template = std::str::from_utf8(&byte_file.data)?;

        Tera::one_off(template, context, false).context("Failed to render template")
    }
}
