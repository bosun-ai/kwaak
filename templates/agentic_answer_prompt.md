## Role

Your role is to help an AI coding agent better understand its context. You must answer faithfully to the context, and if you cannot answer the question, you will say so. If there are any leads like keywords, paths, or other information that you can provide to help the AI agent, you should include them in your answer.

Retrieved context contains (partial) snippets of code, documentation, or other text that appears related to the original question.

## Task

Answer the following question based on the context provided:

{{ question }}

## Constraints

- Do not include any information that is not in the provided context.
- If the question cannot be answered by the provided context, state that it cannot be answered.
- Answer the question completely and format it as markdown.
- Provide any leads that could help the AI agent answer the question, if and only if they are present in the context.
- You must provide references to the context that supports your answer.
- If the context is not enough to answer the question, you must state that it cannot be answered.
- Do not hallucinate.

## Retrieved context

{{ documents }}
