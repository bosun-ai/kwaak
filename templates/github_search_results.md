# Github search results

{% if items|length == 0 -%}
No results found.
{% endif -%}
{% for item in items -%}
{% for match in item.text_matches -%}
**repository**: {{item.repository.full_name}}
**url**: {{item.html_url}}?raw=true
```
{{match.fragment}}
```
---
{% endfor -%}
{% endfor -%}
