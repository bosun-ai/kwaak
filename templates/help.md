You can use the following slash commands to interact with kwaak and the agent:

{% for command, description in commands -%}
**/{{command}}** - {{description}}
{% endfor %}

Additionally, you can use the following keybindings to navigate the interface:

"Page Up/Down - Scroll"
"End - Scroll to end"
"^s - Send a message to the agent"
"^x - Stop agent"
"^n - New chat"
"^q - Quit"

If you encounter any issues, reach out to us on github or discord, we're happy to help!
