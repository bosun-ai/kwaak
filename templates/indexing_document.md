{% if metadata.path %}
The following snippet is part of the file `{{ metadata.path }}`
{% endif %}

````
{{ content }}
````
