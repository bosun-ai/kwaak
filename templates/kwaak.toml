# This is a a prefilled template for `kwaak`
#
# Several assumptions and defaults have been filled in. For proper usage, please customize the values to your needs.
project_name = "{{ project_name }}"
language = "{{ language }}"

## If you are using OpenAI, set the api key here
#openai_api_key = "env:OPENAI_API_KEY"

## Optional: Connect kwaak to github to create PRs, search code, and automatically push to a  remote
#github_api_key = "env:GITHUB_TOKEN"

## Optional: Connect kwaak to tavily to enable it to search the web
#tavily_api_key = "env:TAVILY_API_KEY"

## Commands the agent uses for tools
## Test and coverage are required. Kwaak expects coverage to run the tests as well and echo both the test and coverage results to stdout
## At the moment, the format of the output does not matter.
[commands]
# Example: test = "cargo test --no-fail-fast --color=never"
test = "REQUIRED"
# Example: coverage = "cargo llvm-cov --no-clean --summary-only"
coverage = "REQUIRED"
## Optional: Lint and fix command. This command is run after each completion cycle, before committing the code.
## Recommended to use, as it avoids the LLM getting distracted by linting issues
# Example: lint_and_fix = "cargo clippy --fix --allow-dirty --allow-staged && cargo fmt"

## Git and GitHub configuration
#
## Kwaak can create and update PRs on Github, search github code, and interact with the git repository. This requires a github token.
## If you leave the token empty, kwaak will not create PRs.
[git]
owner = "{{ github.owner }}"
repository = "{{ github.repository }}"
main_branch = "{{ github.main_branch }}"
## The token is expected to have read and write access for git, and the ability to create and update pull requests.
# token = "env:KWAAK_GITHUB_TOKEN"

## Kwaak uses different LLMs for different tasks. As a rule of thumb, tasks that happen often (like indexing, summarizing) require a small, fast model
## and tasks that happen less often (like completion) can use a larger, more accurate model.
#
## You can overwrite the api key per kind of task if needed.
[llm.indexing]
#api_key = "env:KWAAK_OPENAI_API_KEY"
provider = "OpenAI"
prompt_model = "gpt-4o-mini"

[llm.query]
#api_key = "env:KWAAK_OPENAI_API_KEY"
provider = "OpenAI"
prompt_model = "gpt-4o"

[llm.embedding]
#api_key = "env:KWAAK_OPENAI_API_KEY"
provider = "OpenAI"
embedding_model = "text-embedding-3-large"

## Docker configuration
## kwaak requires a Dockerfile for the tool execution environment.
## Besides the dependencies to run the code, there are several additional dependencies:
## - `git` for interacting with the codebase
## - `rg` (ripgrep) for searching the codebase
## - `fd` (fd) for effective file searching
##
## In the future, an executor is planned that does not have these dependencies, but for now, they are required.
##
## If your project already has a Dockerfile and you want to keep it clean, you can specify a different file to use.
[docker]
dockerfile = "Dockerfile"
