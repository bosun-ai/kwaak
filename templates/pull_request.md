{{description}}

---

_This pull request was created by [kwaak](https://github.com/bosun-ai/kwaak), a free, open-source, autonomous coding agent tool. Pull requests are tracked in bosun-ai/kwaak#48_

<details>
<summary>Message History</summary>

{% for message in messages -%}
<details>
  <summary>{{message.role}}</summary>

```markdown
{{message.content}}
```
</details>
{% if message.role is containing("Assistant") -%}

---
{% endif -%}
{% endfor -%}

</details>
