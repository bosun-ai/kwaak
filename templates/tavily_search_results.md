{% if answer -%}
{{ answer }}
{% else -%}
{% for item in results -%}
title: {{ item.title }}
**url**: {{ item.url }}
{{ item.content }}
---

{% endfor -%}
{% endif -%}

{% if answer %}

## Relevant urls

{% for item in results -%}
{{ item.title }}
{{ item.url }}

{% endfor -%}
{% endif %}

{% if follow_up_questions -%}
## Follow up questions

{% for item in follow_up_questions -%}
- {{ item }}
{% endfor -%}
{% endif -%}
