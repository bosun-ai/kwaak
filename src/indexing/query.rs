use anyhow::Result;
use indoc::formatdoc;
use swiftide::traits::EvaluateQuery;
use swiftide::{
    query::{
        self, answers, query_transformers, search_strategies::SimilaritySingleEmbedding, states,
        Query,
    },
    traits::{EmbeddingModel, Persist, Retrieve, SimplePrompt},
};

use crate::{repository::Repository, templates::Templates, util::strip_markdown_tags};

#[tracing::instrument(skip_all, err)]
pub async fn query<S>(
    repository: &Repository,
    storage: &S,
    query: impl AsRef<str>,
) -> Result<String>
where
    S: Retrieve<SimilaritySingleEmbedding> + Persist + Clone + 'static,
{
    // Ensure the table exists to avoid dumb errors
    let _ = storage.setup().await;

    let answer = build_query_pipeline(repository, storage, None)?
        .query(query.as_ref())
        .await?
        .answer()
        .to_string();
    Ok(strip_markdown_tags(&answer))
}

/// Builds a query pipeline
///
/// # Panics
///
/// Should be infallible
pub fn build_query_pipeline<'b, S>(
    repository: &Repository,
    storage: &S,
    evaluator: Option<Box<dyn EvaluateQuery>>,
) -> Result<query::Pipeline<'b, SimilaritySingleEmbedding, states::Answered>>
where
    S: Retrieve<SimilaritySingleEmbedding> + Clone + 'static,
{
    let backoff = repository.config().backoff;
    let query_provider: Box<dyn SimplePrompt> = repository
        .config()
        .query_provider()
        .get_simple_prompt_model(backoff)?;
    let embedding_provider: Box<dyn EmbeddingModel> = repository
        .config()
        .embedding_provider()
        .get_embedding_model(backoff)?;

    let search_strategy: SimilaritySingleEmbedding<()> = SimilaritySingleEmbedding::default()
        .with_top_k(30)
        .to_owned();

    let prompt_template = Templates::from_file("agentic_answer_prompt.md")?;
    let document_template = Templates::from_file("indexing_document.md")?;

    // NOTE: Changed a lot to tailor it for agentic flows, might be worth upstreaming
    // Simple takes the retrieved documents, formats them with a template, then throws it into a
    // prompt with to answer the original question properly. It's really simple.
    let simple = answers::Simple::builder()
        .client(query_provider.clone())
        .prompt_template(prompt_template.into())
        .document_template(document_template)
        .build()
        .expect("infallible");

    let language = repository.config().language.to_string();
    let project = repository.config().project_name.clone();

    let mut pipeline = query::Pipeline::from_search_strategy(search_strategy);

    if let Some(evaluator) = evaluator {
        pipeline = pipeline.evaluate_with(evaluator);
    }

    Ok(pipeline
        .then_transform_query(move |mut query: Query<states::Pending>| {
            let current = query.current();
            query.transformed_query(formatdoc! {"
                {current}

                The project is written in: {language}
                The project is called: {project}

            "});

            Ok(query)
        })
        .then_transform_query(query_transformers::GenerateSubquestions::from_client(
            query_provider.clone(),
        ))
        .then_transform_query(query_transformers::Embed::from_client(
            embedding_provider.clone(),
        ))
        .then_retrieve(storage.clone())
        // .then_transform_response(response_transformers::Summary::from_client(
        //     query_provider.clone(),
        // ))
        .then_answer(simple))
}

#[cfg(test)]
mod tests {

    #[test]
    fn test_rendering_document() {
        use insta::assert_snapshot;
        use swiftide::{indexing::Metadata, query::Document};
        use tera::Context;

        use crate::templates::Templates;

        let document = Document::new(
            "This is a test document",
            Some(Metadata::from([
                ("path", serde_json::Value::from("my file")),
                ("soups", serde_json::Value::from(["tomato", "snert"])),
                ("empty", serde_json::Value::from("")),
            ])),
        );

        assert_snapshot!(Templates::render(
            "indexing_document.md",
            &Context::from_serialize(document).unwrap(),
        )
        .unwrap());
    }
}
