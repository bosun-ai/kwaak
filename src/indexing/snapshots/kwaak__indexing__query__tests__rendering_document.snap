---
source: src/indexing/query.rs
expression: "Templates::render(\"indexing_document.md\",\n&Context::from_serialize(document).unwrap(),).unwrap()"
---
The following snippet is part of the file `my file`


````
This is a test document
````
